bin/faucet/src/faucet/mod.rs

```diff
@@ -292,26 +308,61 @@ impl Faucet {
         Ok(())
     }
 
-    /// Fully handles a single request _without_ changing local state.
+    /// Fully handles a batch of requests _without_ changing local state.
     ///
     /// Caller should update the local state based on the returned result.
-    async fn handle_request(
+    async fn handle_request_batch(
         &self,
-        request: MintRequest,
-        rng: impl FeltRng,
+        requests: &[MintRequest],
+        rng: &mut impl FeltRng,
         rpc_client: &mut RpcClient,
-    ) -> MintResult<(AccountDelta, BlockNumber, Note)> {
+    ) -> MintResult<(AccountDelta, BlockNumber, Vec<Note>)> {
+        let (executed_transaction, notes) = self.create_batch_transaction(requests, rng)?;
+        let account_delta = executed_transaction.account_delta().clone();
+        let tx = Self::prove_transaction(executed_transaction)?;
+        let block_number = self.submit_transaction(tx, rpc_client).await?;
+
+        Ok((account_delta, block_number, notes))
+    }
+
+    /// Creates and executes a transaction that outputs the requested notes.
+    fn create_batch_transaction(
+        &self,
+        requests: &[MintRequest],
+        rng: &mut impl FeltRng,
+    ) -> MintResult<(ExecutedTransaction, Vec<Note>)> {
         // Generate the payment note and compile it into our transaction arguments.
-        let p2id_note = P2IdNote::build(self.faucet_id(), &request, rng)?;
-        let tx_args = p2id_note.compile()?;
+        let p2id_notes: Vec<P2IdNote> = requests
+            .iter()
+            .filter_map(|request| {
+                let note = P2IdNote::build(self.faucet_id(), request, rng);
+                match note {
+                    Ok(note) => Some(note),
+                    Err(err) => {
+                        tracing::error!(request.account_id=%request.account_id, ?err, "failed to build note");
+                        None
+                    }
+                }
+        }).collect();
```

Comment:
I wonder if `P2IdNote` should be renmaed to `P2IdNotes` and take a list of requests instead, and then this piece of can be part of the building process. It could also have a single `into_inner()` that returns the `Vec<Note>` directly and then you can avoid collecting into a different vector on line 347.